packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts

```diff
@@ -1,20 +1,358 @@
+            .integerValue()
+            .toNumber();
+        const outputFees = [fee, fee, fee];
+        // NOTE: ids can be the same for all fake samples
+        const id = sampleToId(ERC20BridgeSource.Native, idx);
+        const ids = [id, id, id];
+
+        const serializedPath: SerializedPath = {
+            ids,
+            inputs,
+            outputs,
+            outputFees,
+        };
+
+        samplesAndNativeOrdersWithResults.push([nativeOrder]);
+        serializedPaths.push(serializedPath);
+    }
+
+    const rustArgs: OptimizerCapture = {
+        side,
+        targetInput: input.toNumber(),
+        pathsIn: serializedPaths,
+    };
+
+    let allSourcesRustRoute: number[] = [];
+    if (rustArgs.pathsIn.length > 0) {
+        const before = performance.now();
+        allSourcesRustRoute = route(rustArgs, RUST_ROUTER_NUM_SAMPLES);
+        DEFAULT_INFO_LOGGER(
+            { router: 'neon-router', performanceMs: performance.now() - before, type: 'real' },
+            'Rust router real routing performance',
+        );
+    }
+
+    const routesAndSamples = _.zip(allSourcesRustRoute, samplesAndNativeOrdersWithResults);
+
+    const adjustedFills: Fill[] = [];
+    const totalRoutedAmount = BigNumber.sum(...allSourcesRustRoute);
+
+    const scale = input.dividedBy(totalRoutedAmount);
+    for (const [routeInput, routeSamplesAndNativeOrders] of routesAndSamples) {
+        if (!routeInput || !routeSamplesAndNativeOrders) {
+            continue;
+        }
+        // TODO(kimpers): [TKR-241] amounts are sometimes clipped in the router due to precisions loss for number/f64
+        // we can work around it by scaling it and rounding up. However now we end up with a total amount of a couple base units too much
+        const rustInput = BigNumber.min(
+            new BigNumber(routeInput).multipliedBy(scale).integerValue(BigNumber.ROUND_CEIL),
+            input,
+        );
+
+        const current = routeSamplesAndNativeOrders[routeSamplesAndNativeOrders.length - 1];
+        if (!isDexSample(current)) {
+            const nativeFill = nativeOrdersToFills(
+                side,
+                [current],
+                rustInput,
+                opts.outputAmountPerEth,
+                opts.inputAmountPerEth,
+                fees,
+            )[0];
+            // NOTE: For Limit/RFQ orders we are done here. No need to scale output
+            adjustedFills.push(nativeFill);
+            continue;
+        }
+
+        // NOTE: For DexSamples only
+        let fill = createFill(current);
+        const routeSamples = routeSamplesAndNativeOrders as Array<DexSample<FillData>>;
+        // Descend to approach a closer fill for fillData which may not be consistent
+        // throughout the path (UniswapV3) and for a closer guesstimate at
+        // gas used
+        for (let k = routeSamples.length - 1; k >= 0; k--) {
```

Comment:
I feel like so much of this would break unexpectedly or silently if there was a bug in the router because there's a lot of assumptions going on with the consistency of the router's outputs. Maybe we could pepper in asserts?